### ISSUE TITLE
onClosing doesn't work if Popout is given a URL

### ISSUE THREAD

**Issue by tocttou** [JakeGinnivan/react-popout#22]
onClosing is only triggered if you close it before the URL has loaded. This is probably because the component loaded inside is gone.

**Comment by guvenkemal** [comment]
I am also having this issue.

**Comment by renholm** [comment]
Same problem here. Could it be that the navigation even causes the function to be cleared?

**Comment by JeffLabonte** [comment]
I am having the same problem here.

**Comment by ghost** [comment]
Same.

**Comment by erisnuts** [comment]
Did somebody handeled this issue ?

**Comment by marcellodesales** [comment]
Anyone? Need this... :(

**Comment by marchaos** [comment]
Can you re-test this with the latest version? Should work fine with the latest version.

**Comment by jaydlawrence** [comment]
I am on `react-popout@1.0.1` and I am having this issue.
Chrome Version 61.0.3163.100 (Official Build) (64-bit) on Ubuntu
Does the same on Firefox 56.0 (64-bit) too.

In the code below, it never calls the console.log, but it does if I remove the url prop.

```javascript
renderPopoutWindow = (renderPopoutWindow) => {
    if (!renderPopoutWindow) return null

    const justWork = () => {
      console.log('It worked!!!!')
    }

    return (
      <Popout
        url='https://google.com'
        title='Google'
        onClosing={justWork}
      >
        <div>Popped out content!</div>
      </Popout>
    )
  }
```

**Comment by marchaos** [comment]
Hi @jaydlawrence. How are you closing the window? Just with the window's close button?

**Comment by jaydlawrence** [comment]
Yes. But programmatically closing the window from within, using window.top.close() in a script inside the popup window had the same effect.

**Comment by jaydlawrence** [comment]
I have had a look into this further.

The 'beforeUnload' event listener definitely doesn't exist after going to a URL.
I wonder if it is that the event listeners are bound to the DOM initially which changes when the window loads a new URL?

I made this quick hacky test and looks like the window.closed works even after loading a different page and after redirects.

``` javascript
handleClose = () => {
    console.log('here')
  }

  wait = (windowHandle) => {
    console.log(windowHandle.closed)
    setTimeout(this.wait, 1000, windowHandle)
  }

  hackyPopout = () => {
    const popOutWindow = window.open(
      "https://google.com",
      "_blank",
      "toolbar=yes,scrollbars=yes,resizable=yes,top=500,left=500,width=400,height=400"
    )
    popOutWindow.addEventListener(
      'beforeunload',
      this.handleClose
    )
    this.wait(popOutWindow)
  }
```

Setting the URL to `null` makes the `beforeUnload` listener work.

**Comment by rabashani** [comment]
did you get answer on this one?

**Comment by jaydlawrence** [comment]
@rabashani, no. I ended up making my own version using window.closed

**Comment by marcellodesales** [comment]
@jaydlawrence mind sharing your solution somewhere?

**Comment by tocttou** [comment]
I see that this issue is still bugging many people. Here is how this can be solved easily using "Portals" in React 16. https://hackernoon.com/using-a-react-16-portal-to-do-something-cool-2a2d627b0202

**Comment by eragon512** [comment]
hey, I am still facing this issue as well, any way to fix it?

**Comment by willsmanley** [comment]
I am making a fork that uses an interval to check every 250ms to see if the window is closed. Even with this check, trying to unmount the node doesn't work if a URL is passed.
```
unmountComponentAtNode(...): Target container is not a DOM element.
```

So you can just call your onClosing() method without using the unmountComponentAtNode() function.

**Comment by willsmanley** [comment]
This fork works really nicely for me:
```
openPopoutWindow(ownerWindow) {
        const popoutWindow = ownerWindow.open(this.props.url, this.props.name || this.props.title, this.createOptions(ownerWindow));
        if (!popoutWindow) {
            this.props.onError();
            return;
        }
        this.setState({ popoutWindow });
        
        popoutWindow.addEventListener('load', this.popoutWindowLoaded);
        popoutWindow.addEventListener('beforeunload', this.popoutWindowUnloading);

        // If they have no specified a URL, then we need to forcefully call popoutWindowLoaded()
        if (this.props.url === ABOUT_BLANK) {
            popoutWindow.document.readyState === 'complete' && this.popoutWindowLoaded(popoutWindow);
        }else{
            // If a URL prop is passed, check for closure of the cross-domain popup window.
            this.checkForPopoutWindowClosure(popoutWindow);
        }
    }

    checkForPopoutWindowClosure(popoutWindow) {
        let interval = setInterval(()=>{
            if(popoutWindow.closed){
                clearInterval(interval); // stop looping
                this.props.onClosing && this.props.onClosing(); // call your onClosing method, if any
            }
        },500);
    }
```

I don't have time to submit a PR right now but perhaps you could check on this @JakeGinnivan?

**Comment by willsmanley** [comment]
I went ahead and added a PR for this solution from my [fork](https://github.com/willsmanley/react-popout)

#56

**Comment by whns** [comment]
@famouspotatoes Thanks, this approach worked for me also!
@JakeGinnivan @marchaos Seems that this is still an active issue with 1.0.2.

**Comment by marchaos** [comment]
You need 1.0.3 and you need to add an onClosing callback.

Cheers
Marc

**Comment by marchaos** [comment]
Fixed in 1.0.3. You need to add an onClosing callback to the window to utilise the fix.
